modules/navier_stokes/src/actions/NSFVAction.C

```diff
@@ -0,0 +1,2259 @@
+  InputParameters params = _factory.getValidParams(kernel_type);
+  params.set<NonlinearVariableName>("variable") = NS::T_fluid;
+  params.set<MooseFunctorName>(NS::T_fluid) = NS::T_fluid;
+  params.set<bool>("is_solid") = false;
+
+  for (unsigned int block_i = 0; block_i < num_used_blocks; ++block_i)
+  {
+    std::string block_name = "";
+    if (num_convection_blocks)
+    {
+      params.set<std::vector<SubdomainName>>("block") = {_ambient_convection_blocks[block_i]};
+      block_name = _ambient_convection_blocks[block_i];
+    }
+    else
+      params.set<std::vector<SubdomainName>>("block") = _blocks;
+
+    params.set<MooseFunctorName>("h_solid_fluid") = _ambient_convection_alpha[block_i];
+    params.set<MooseFunctorName>(NS::T_solid) = _ambient_temperature[block_i];
+
+    _problem->addFVKernel(kernel_type, "ambient_convection_" + block_name, params);
+  }
+}
+
+void
+NSFVAction::addINSEnergyExternalHeatSource()
+{
+  const std::string kernel_type = "FVCoupledForce";
+  InputParameters params = _factory.getValidParams(kernel_type);
+  params.set<NonlinearVariableName>("variable") = NS::T_fluid;
+  params.set<std::vector<SubdomainName>>("block") = _blocks;
+  params.set<CoupledName>("v") = getParam<CoupledName>("external_heat_source");
+  params.set<Real>("coef") = getParam<Real>("external_heat_source_coeff");
+
+  _problem->addFVKernel(kernel_type, "external_heat_source", params);
+}
+
+void
+NSFVAction::addScalarAdvectionKernels()
+{
+  for (const auto & vname : _passive_scalar_names)
+  {
+    const std::string kernel_type = "INSFVScalarFieldAdvection";
+    InputParameters params = _factory.getValidParams(kernel_type);
+    params.set<NonlinearVariableName>("variable") = vname;
+    params.set<MooseEnum>("velocity_interp_method") = "rc";
+    params.set<MooseEnum>("advected_interp_method") = _passive_scalar_advection_interpolation;
+
+    if (_porous_medium_treatment)
+      params.set<UserObjectName>("rhie_chow_user_object") = "pins_rhie_chow_interpolator";
+    else
+      params.set<UserObjectName>("rhie_chow_user_object") = "ins_rhie_chow_interpolator";
+
+    _problem->addFVKernel(kernel_type, "ins_" + vname + "_advection", params);
+  }
+}
+
+void
+NSFVAction::addScalarDiffusionKernels()
+{
+  for (unsigned int name_i = 0; name_i < _passive_scalar_names.size(); ++name_i)
+  {
+    const std::string kernel_type = "FVDiffusion";
+    InputParameters params = _factory.getValidParams(kernel_type);
+    params.set<NonlinearVariableName>("variable") = _passive_scalar_names[name_i];
+    params.set<std::vector<SubdomainName>>("block") = _blocks;
+    params.set<MooseFunctorName>("coeff") = _passive_scalar_diffusivity[name_i];
+
+    _problem->addFVKernel(
+        kernel_type, "ins_" + _passive_scalar_names[name_i] + "_diffusion", params);
+  }
+}
+
+void
+NSFVAction::addScalarSourceKernels()
+{
+  for (unsigned int name_i = 0; name_i < _passive_scalar_names.size(); ++name_i)
+  {
+    const std::string kernel_type = "FVBodyForce";
```

Comment:
we need FVCoupledForce to be able to pass a fission source auxvariable as well as a coeff, different for each precursor

Reply:
Added!